Created tokenization & authentication functions

JonasRSV opened the pull request:
### Issues: #13 

### Test Plan
Have not made any comprehensive tests while running everything only one test for the actual tokenization and authentication functions.
### Description
Implements jjwt to create and authenticate JWTs

Commit:
Created tokenization & authentication functions

jvikstrom commented on user-service/src/main/java/sjchat/users/UserAuthentication.java:
```diff
@@ -0,0 +1,126 @@
+    private static final UserAuthentication INSTANCE = new UserAuthentication();
+  }
+
+  public UserAuthentication checkUserExists(String username, String password) throws AuthenticationException {
```
It feels like this should be a part of the UserService class.

jvikstrom commented on user-service/src/main/java/sjchat/users/UserAuthentication.java:
```diff
@@ -0,0 +1,126 @@
+    return this;
+  }
+
+  public UserAuthentication checkUsernameExists(String username) throws AuthenticationException {
```
It feels like this should be a part of the UserService class

JonasRSV replied:
Indeed & The UserUpdate static class should probably be to.. Unless i rename the entire class to something more fitting? I kind of like the flow of having that functionality in there. It made the code in UserAuth look more clean in my opinion. But i can change it.

jvikstrom commented on proto/user-service.proto:
```diff
@@ -52,6 +64,20 @@ message UpdateUserResponse {
+}
+
+message LoginRequest {
+    string username = 1;
```
Why does LoginRequest and LogoutRequest require both a token and a username?
Shouldn't it only be the token that is required, and then the token tells us what userid it belongs to?

JonasRSV replied:
LoginRequest only requires a Username & password? 

AuthRequest needs a token & username to verify that the token comes from the right user. Or that was the idea behind it.

JonasRSV replied:
Logout requires a token to verify that the actual user is trying to logout, can remove that if you think it's unneccecary

jvikstrom replied:
Oh, I only meant AuthRequest, don't know why I wrote LoginRequest to.
The token is obviously needed but I don't think the username should be.

Shouldn't  the token contain the id of the user that the token belongs to and that is how you do auth checking? I don't think having the username as an additional parameter makes this any more secure then normal jwts.

JonasRSV replied:
The token does contains the users username "amongst other things" and we verify the tokens content with the username of the user which the token belongs to. Otherwise we'll have to do that check somewhere else to make sure that the owner of the token is the rightful one. 

Perhaps i've misintepreted the question? 

But the reason why Auth request takes a username aswell is to verify that the token comes from the correct user. Perhaps that check is done somewhere else i'm not aware of, in that case i'll remove the username field.

jvikstrom replied:
But isn't that check (token.username == request.username) really not necessary?

If the token has a claim for username "superawesomeuser", the token belongs to "superawesomeuser" and anything we're doing that requires a token should use the "superawesomeuser" user.
The only way you should be able to get a token is by sending a request to the "/login" with username and password. That token is never given away any other way.

Also the username in the token can easily be retrieved from the token. Simply decode the message from base64 into ascii (or whatever).  So that check doesn't really add any security

JonasRSV replied:
Fair Points! :) I've made the changes you suggested

jvikstrom commented on proto/user-service.proto:
```diff
@@ -31,10 +34,19 @@ message UserDataRequest {
-    string id = 1;
+
+message LoginResponse {
+    bool authenticated = 1;
```
There is apparently a oneof "attribute" you could probably use here.

https://developers.google.com/protocol-buffers/docs/proto3#oneof

JonasRSV replied:
Added it!

Commit:
Update Pull Request

added oneOf
removed username field in authRequest
updated AuthFunction so that it nolonger requires username
